### PR TITLE
updates implementation of memory-optimal model loading to eliminate excess consumption

### PR DESCRIPTION
**Pull request by RobotSail**
# Memory-Efficient Distributed Model Loading for OSFT and SFT

## Problem Statement

The previous implementation had a critical memory inefficiency during model initialization in distributed training: **every GPU process loaded a complete copy of the model into CPU memory**. For large models (e.g., 20B+ parameters) trained on 8+ GPUs, this resulted in:
- **8x redundant CPU memory consumption** (one full model copy per GPU)
- **Out-of-memory (OOM) errors** on systems with limited CPU RAM
- **Slower initialization** due to redundant I/O and deserialization

This was particularly problematic for OSFT training, which requires additional memory overhead for SVD decomposition and subspace tracking.

## Solution Overview

This PR implements a **memory-efficient distributed loading strategy** that eliminates redundant CPU memory usage:

- **Rank 0**: Loads the full model to CPU (required for parameter distribution)
- **All other ranks**: Create lightweight meta-device models (no actual memory allocation)
- **After FSDP2 sharding**: State dict is efficiently broadcast from rank 0 to all ranks

This reduces peak CPU memory from `N × model_size` to `~1 × model_size` for N-GPU training.

### Three-Phase Initialization Pipeline

The refactor introduces a clean three-phase pipeline that works for both SFT and OSFT:

**Phase 1: `prepare_model_for_fsdp2()`** (src/mini_trainer/setup_model_for_training.py:278-348)
- Detects lazy initialization mode (SFT vs OSFT)
- Extracts state dict from rank 0 only
- Materializes buffers from meta device
- Returns context for subsequent phases

**Phase 2: `wrap_fsdp2()`** (src/mini_trainer/setup_model_for_training.py:351-428)
- Pure FSDP2 wrapping with activation checkpointing
- Applies per-block sharding following TorchTitan's approach
- No state dict handling (cleanly separated from Phase 1/3)

**Phase 3: `finalize_model_initialization()`** (src/mini_trainer/setup_model_for_training.py:431-514)
- **SFT**: Broadcasts state dict via `set_model_state_dict()`
- **OSFT**: Distributes non-OSFT params, then computes distributed SVD
- Sanitizes meta tensor attributes
- Cleans up lazy init flags

### Key Changes

**1. OSFT Distributed Setup** (src/mini_trainer/setup_model_for_training.py:617-703)
- New `setup_osft_model_distributed()` function
- Rank 0 loads to CPU with `fsdp2_lazy_init=True`
- Non-zero ranks create meta device model (src/mini_trainer/osft_utils.py)
- Deferred SVD computation until after FSDP2 wrapping

**2. SFT Distributed Setup** (src/mini_trainer/setup_model_for_training.py:706-814)
- New `setup_sft_model_distributed()` function
- Extends memory-efficient loading to SFT (previously OSFT-only)
- Rank 0 loads model and extracts config/state dict/buffers
- All ranks create meta device model from config

**3. Lazy Init Mode Tagging** (src/mini_trainer/fsdp2_lazy_init.py)
- New module for tracking initialization mode (`sft` vs `osft`)
- `set_fsdp2_lazy_init_mode()` / `get_fsdp2_lazy_init_mode()` utilities
- Enables phase 1 to detect which finalization path to use

**4. OSFT Refactoring** (src/mini_trainer/osft_utils.py)
- Extracted `_build_osft_kwargs()` for cleaner arg handling
- New `eject_og_state_dict()` method for lazy init (osft_utils.py:899-917)
- `post_fsdp2_wrap_synchronize_state_dict_across_procs()` for non-OSFT param distribution (osft_utils.py:1015-1056)
- `compute_distributed_svd()` for distributed SVD computation (osft_utils.py:1058-1113)
- Better separation of concerns between initialization and training

### Deprecated Parameters
- **`osft_memory_efficient_init`**: Now deprecated (src/mini_trainer/training_types.py:145-149, api_train.py:81-92)
  - Memory-efficient initialization is **automatically enabled** when `torch.distributed.is_initialized()` returns `True`
  - Flag has no effect and will be removed in v0.5.0
  - Updated CHANGELOG.md with deprecation notice

### New TorchrunArgs Support
- Added `master_addr` and `master_port` for static rendezvous (training_types.py:37-38)
- Auto-selects `--rdzv-backend=static` when using `master_addr` (api_train.py:110-119)
- Fixes compatibility with standard torchrun workflows

### Fixed Torchrun Argument Formatting
- Changed from underscore to hyphen format (e.g., `--nproc-per-node` instead of `--nproc_per_node`) (api_train.py:102-108)
- Matches torchrun's expected argument format

## Additional Improvements

**1. Model Support**
- Added Qwen3 support (src/mini_trainer/utils.py, various test files)
- Fixed Granite 4 attention kernel initialization (src/mini_trainer/gpt_oss_utils.py:13-47)

**2. Robustness**
- Added try/finally blocks to prevent memory leakage (src/mini_trainer/osft_utils.py)
- Fixed dtype mismatch when loading GPT-OSS state dicts (src/mini_trainer/gpt_oss_utils.py)
- Generic meta tensor attribute sanitization (setup_model_for_training.py:63-159)
- Better handling of hidden tensor attributes in distributed OSFT

**3. Testing**
- New comprehensive OSFT orthogonalization regression test (regression_tests/test_osft_orthogonalization.py)
- 500+ line test suite for gradient/parameter orthogonality checks
- Updated integration tests for new distributed loading path
- Added orthogonality test utilities (tests/test_utils/orthogonality.py)

**4. User Experience**
- Fixed progress bar epoch counter (now 1-indexed) (src/mini_trainer/async_structured_logger.py:124-145)
- CI optimizations: linting/formatting now only run on changed files (.github/workflows/pr-tests.yml)
- Code cleanup: removed dead code, simplified FSDP2-related attributes

## Files Changed (21 files, +3182/-668 lines)

**Core Implementation:**
- `src/mini_trainer/setup_model_for_training.py`: Main refactor (+835 lines)
- `src/mini_trainer/osft_utils.py`: OSFT distributed loading (+1273 lines refactored)
- `src/mini_trainer/fsdp2_lazy_init.py`: New lazy init mode tracking (+34 lines)

**API & Configuration:**
- `src/mini_trainer/api_train.py`: Torchrun args, deprecation warnings (+53/-21 lines)
- `src/mini_trainer/training_types.py`: API updates (+33/-0 lines)
- `src/mini_trainer/train.py`: Integration with new pipeline (+28/-0 lines)

**Testing:**
- `regression_tests/test_osft_orthogonalization.py`: New comprehensive test (+501 lines)
- `tests/test_osft.py`: Updated OSFT tests (+394 lines)
- `tests/test_model_initialization.py`: Distributed loading tests (+167 lines)
- `tests/test_utils/orthogonality.py`: New test utilities (+269 lines)

**Documentation:**
- `CHANGELOG.md`: Deprecation notice and feature documentation (+14 lines)

## Testing

✅ All existing tests pass with updated initialization path
✅ New regression test validates OSFT orthogonality over 100 training steps
✅ Integration tests verify distributed loading for both SFT and OSFT
✅ Manually tested on multi-GPU setups (2-8 GPUs) with 7B-20B models

## Migration Guide

**For users currently using `osft_memory_efficient_init=True`:**
```python
# Before (still works but deprecated)
train_args = TrainingArgs(
    model_name="openai/gpt-oss-20b",
    osft_memory_efficient_init=True,  # ⚠️ Deprecated
    ...
)

# After (recommended)
train_args = TrainingArgs(
    model_name="openai/gpt-oss-20b",
    # osft_memory_efficient_init=True,  # Remove this line
    ...
)
```

Memory-efficient initialization now happens **automatically** in distributed mode. No code changes required!

## Summary by CodeRabbit

* **New Features**
  * Three‑phase distributed initialization (prepare → wrap → finalize) with FSDP2 lazy‑init and OSFT/SFT memory‑efficient loading; new control process group for coordination.

* **Deprecated**
  * osft_memory_efficient_init flag deprecated, now ignored at runtime and emits deprecation warnings (removal in v0.5.0).

* **Bug Fixes**
  * Clearer dtype/gradient validation messages and improved distributed synchronization and weight handling.

* **Tests**
  * End‑to‑end OSFT integration and lazy‑init tokenizer alignment tests added; numerous test updates.

* **Documentation**
  * New distributed initialization guide describing the three‑phase flow.

* **Chores**
  * CI now lints/formats only changed Python files; various tooling and ignore updates.

### CODE REVIEW

File: src/mini_trainer/osft_utils.py [hunk 1]
```diff
@@ -968,53 +1033,365 @@ def from_pretrained(
...
+                model_state_dict=finalized_sd,
+                options=StateDictOptions(
+                    full_state_dict=True,
+                    broadcast_from_rank0=True,
+                    strict=False,
+                )
+            )
+            
+            log_rank_0("✅ [compute_distributed_svd] Distributed SVD computation complete!")
+
```
[the remark this answers is not in the document]
Comment (RobotSail) [reply src/mini_trainer/osft_utils.py]: Here are my responses to your points:

1. You're wrong about this, it's a new syntax introduced in Python3.9
2. True, I'm fixing this locally
3. I fixed the annotation from earlier

File: src/mini_trainer/osft_utils.py [hunk 2]
```diff
@@ -730,85 +682,97 @@ def _load_model_memory_efficient(
...
+    log_rank_0("instantiating OSFT model on meta device")
+    extra_kwargs = {}
+    if dist.get_rank() == 0:
+        # rank 0 passes the OG state dict to the OSFT model for later access
+        extra_kwargs['lazy_init_og_state_dict'] = state_dict
     
+    # this instantiates the model as we'd expect, except that params and buffers
+    # only load their metadata (dtype, shape) without loading the raw data or creating new random weights
+    with torch.device("meta"):
+        model = actual_osft_cls(
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/osft_utils.py]: Nice!!

File: src/mini_trainer/osft_utils.py [hunk 3]
```diff
@@ -968,53 +1032,422 @@ def from_pretrained(
...
+            for sender_rank in range(dist.get_world_size()):
+                # again, rank 0 doesn't need to send to itself
+                if sender_rank == main_proc_rank:
+                    if is_main_proc:
+                        gathered_results.update(processed_svd_dicts)
+                    continue
+
+                # worker procs send back to main proc
+                mailbox = [None]
+                if sender_rank == current_rank:
```
Comment (NikhilNayak-debug) [inline src/mini_trainer/osft_utils.py]: I assume this works after all the tests you ran, but it is a bit hard to read. I wonder if the code can be made easier to follow here.
